Move the creation of config files to run time

josemauro opened the pull request:
Moving the creation of config files to run time instead installation time. 
- The 'config.py' create a folder 'etc/kytos' and 'kytos.conf' if the file is not found.
- The 'kytosd.py' create 'var/run' and  folders.

josemauro committed:
Moving the creation of config files to run time instead installation time

josemauro committed:
Moving the creation of config files to run time instead installation time

cmagnobarbosa commented on kytos/core/config.py:
```diff
@@ -125,18 +130,55 @@ def parse_args(self):
+                os.makedirs(directory)
+
+        for path in templates:
+            with open(path, 'r', encoding='utf-8') as src_file:
```
Running this modification, the error FileNotFound is displayed.

The error message:

FileNotFoundError: [Errno 2] No such file or directory: 'etc/kytos/kytos.conf.template'

To reproduce this error:

- Create a new virtual environment.
- Clone this repository: git clone git@github.com:kytos/kytos.git
- Get this modification: git fetch origin refs/pull/1044/head
- make checkout for this new branch: git checkout -b josemauro/fix-issue-1042 FETCH_HEAD
- Install python-openflow, kytos-utils, kytos.
- Run kytos: kytosd -f

gleybersonandrade committed:
Fix template error in installation

An error was raised when kytos was trying to copy template files to installation root folder. This commit changes config.py to solve this problem.

hdiogenes commented on kytos/core/kytosd.py:
```diff
@@ -23,6 +29,18 @@ def in_prompt_tokens(self):
+        pid_folder = os.path.join(BASE_ENV, 'var/run/kytos')
+        os.makedirs(pid_folder, exist_ok=True)
+        if BASE_ENV == '/':  # system install
+            os.chmod(pid_folder, 0o1777)  # permissions like /tmp
```
No need for class + static method when all you're going to use is a function.

hdiogenes commented on kytos/core/config.py:
```diff
@@ -125,18 +131,60 @@ def parse_args(self):
         self.options['daemon'] = self._parse_options(argv)
 
+    @staticmethod
+    def _generate_file_from_template(templates,
```
`_generate_file_from_template` can be moved to a function instead of an `staticmethod`

hdiogenes commented on kytos/core/config.py:
```diff
@@ -125,18 +131,60 @@ def parse_args(self):
+            destination (string): Directory in which the config file will
+                                  be placed.
+        """
+        from jinja2 import Template
```
Move import to top of file.

josemauro committed:
improvements in code suggest by humberto

hdiogenes commented on kytos/core/config.py:
```diff
@@ -147,44 +149,6 @@ def parse_args(self):
-    @staticmethod
-    def _generate_file_from_template(templates,
-                                     destination=Path(__file__).parent,
-                                     **kwargs):
```
It looks like this function name should be in the plural form: `_generate_files_from_templates()`, right?
My suggestion: `_render_config_templates()`

josemauro replied:
I liked the suggestion and already applied it.

hdiogenes commented on kytos/core/kytosd.py:
```diff
@@ -84,6 +98,12 @@ def start_shell(controller=None):
     """Read config and start Kytos in foreground or daemon mode."""
+    # data_files is not enough when installing from PyPI
+    for file in ETC_FILES:
+        shutil.copy2(file, Path(BASE_ENV) / file)
```
What's the purpose of this when ETC_FILES is just an empty list?

josemauro replied:
I'm not sure, but It's something that already were in the project, I'll investigate it to understand. For now, I don't see any reason to keep that.

josemauro replied:
It's possible to check that the ETC_FILES got an empty list in commit c4b1a1a9, and stayed empty since then. I just removed this code and tested the installation. Everything seems to work fine.

hdiogenes commented on setup.py:
```diff
@@ -138,60 +132,56 @@ def run(self):
+#         directories = [os.path.join(BASE_ENV, 'etc/kytos')]
+#         for directory in directories:
+#             if not os.path.exists(directory):
+#                 os.makedirs(directory)
```
The commented lines can be removed.

josemauro replied:
It's done.

hdiogenes commented on setup.py:
```diff
@@ -205,61 +195,19 @@ def run(self):
+#
+#    def run(self):
+#        """Install the package in a developer mode."""
+#        super().run()
```
remove commented code

josemauro replied:
It's done.

hdiogenes commented on kytos/core/kytosd.py:
```diff
@@ -23,6 +27,14 @@ def in_prompt_tokens(self):
+    pid_folder = os.path.join(BASE_ENV, 'var/run/kytos')
+    os.makedirs(pid_folder, exist_ok=True)
+    if BASE_ENV == '/':  # system install
+        os.chmod(pid_folder, 0o1777)  # permissions like /tmp
```
`s/folder/directory/` (or `dir`)

josemauro replied:
I replaced it. Thanks.
